### PR TITLE
[JBPM-9296]  Self contained Audit/History module

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9296

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/jbpm/pull/1731

**Comment by afalhambra** [comment]
@elguardian CI is failing on `HistorySwaggerDecorationTest` test

https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/new-PR-droolsjbpm-integration-master.pullrequests/206/consoleText

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/main/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecoration.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import org.kie.server.remote.rest.common.marker.KieServerEndpoint.EndpointType;
+
+import io.swagger.jaxrs.ext.AbstractSwaggerExtension;
+import io.swagger.jaxrs.ext.SwaggerExtension;
+import io.swagger.models.Operation;
+
+
+public class HistorySwaggerDecoration extends AbstractSwaggerExtension {
+
+    public static final String SWAGGER_EXTENSION_READ_ONLY = "x-readonly";
```
Comment (afalhambra) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/main/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecoration.java:38]: Maybe `x-history` makes more sense here? not sure, wdyt?

```suggestion
    public static final String SWAGGER_EXTENSION_READ_ONLY = "x-history"; 
```

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import org.junit.runner.RunWith;
+import org.kie.server.remote.rest.common.marker.KieServerEndpoint;
+import org.kie.server.remote.rest.common.marker.KieServerEndpoint.EndpointType;
+import org.kie.server.remote.rest.common.marker.KieServerEndpointRequestFilter;
+import org.mockito.InjectMocks;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import io.swagger.models.Operation;
+
+@RunWith(MockitoJUnitRunner.class)
```
Comment (afalhambra) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java:36]: We don't need to use Mockito here. Please remove

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            // nothing
+        }
+    }
+    
+
+
+    
+
+    @InjectMocks
+    private KieServerEndpointRequestFilter filter = new KieServerEndpointRequestFilter();
```
Comment (afalhambra) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java:83]: Not needed. Please remove.

**Review by afalhambra — CHANGES_REQUESTED** [review]
Only minor comments. Some improvements requested in test cases.

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/util/QueryParamFactory.java [hunk 4]
```diff
@@ -102,6 +102,10 @@ public static QueryParam type(String column, Comparable<?> type) {
         return new QueryParam(column, "TYPE", Arrays.asList(type));
     }
 
+    public static QueryParam history() {
+        return new QueryParam("TABLE", "MODE", Arrays.asList("HISTORY"));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/util/QueryParamFactory.java:106]: it does not really matter. I set it to table for convenience.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-case-mgmt/src/main/java/org/kie/server/remote/rest/casemgmt/CaseQueryResource.java [hunk 5]
```diff
@@ -376,10 +378,12 @@ public Response getCaseInstanceDataItems(@javax.ws.rs.core.Context HttpHeaders h
                 });
     }
 
+    @ApiOperation(value = "Queries cases by variables")
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-case-mgmt/src/main/java/org/kie/server/remote/rest/casemgmt/CaseQueryResource.java:381]: done

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/marker/KieServerEndpointRequestFilter.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+@Provider
+public class KieServerEndpointRequestFilter implements ContainerRequestFilter {
+
+    private final List<EndpointType> valid = Arrays.asList(KieServerEndpoint.EndpointType.ALWAYS, KieServerEndpoint.EndpointType.HISTORY);
+
+    @Context
+    private ResourceInfo resourceInfo;
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/marker/KieServerEndpointRequestFilter.java:44]: removed

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/RuntimeDataResource.java [hunk 7]
```diff
@@ -766,10 +768,12 @@ public Response getTasksByVariables(@Context HttpHeaders headers,
         }
     }
 
+    @ApiOperation(value = "Queries processes by variables")
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/RuntimeDataResource.java:771]: done

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/main/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecoration.java (hunk 1 above)
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/main/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecoration.java:38]: not really. At first I thought history was the good one, but all operations are read only for database. I think makes more sense.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java (hunk 2 above)
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java:36]: removed

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java (hunk 3 above)
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-swagger/src/test/java/org/kie/server/remote/rest/swagger/HistorySwaggerDecorationTest.java:83]: removed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RuntimeDataServiceIntegrationTest.java [hunk 8]
```diff
@@ -1866,6 +1868,12 @@ public void testQueryProcessesByVariables() throws Exception {
         assertNotNull(listProcesses);
         listProcesses.stream().forEach(e -> assertEquals(CONTAINER_ID, e.getContainerId()));
 
+        SearchQueryFilterSpec specHistory = new SearchQueryFilterSpec();
+        specHistory.setAttributesQueryParams(list(history(), equalsTo(PROCESS_ATTR_DEPLOYMENT_ID, CONTAINER_ID)));
+        List<ProcessInstanceCustomVars> listHistoryProcesses = queryClient.queryProcessesByVariables(specHistory, 0, 2);
+        assertNotNull(listHistoryProcesses);
+        listHistoryProcesses.stream().forEach(e -> assertEquals(CONTAINER_ID, e.getContainerId()));
+
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/RuntimeDataServiceIntegrationTest.java:1876]: actually we are doing. history is doing the integration test with audit tables already. This is tested at jbpm side. See code coverage.

**Comment by afalhambra** [comment]
Jenkins retest this

**Comment by elguardian** [comment]
Jenkins retest this please.

**Review by afalhambra — APPROVED** [review]
Very good job.

**Pull request merged**